### ISSUE TITLE
[rstream] sync with req and opt sources?

### ISSUE THREAD

**Issue by den1k** [thi-ng/umbrella#41]
I was implementing a file-browser and ran into the following issue:

The file-browser views a list of files and lets the user select on of them. It has `files` and `selectedFile` streams as it's sources.

```javascript
const files = rs.stream()
const selectedFile = rs.stream()
const selectedFileData = selectedFile.subscribe(tx.map(processFile...))

const fileBrowser = rs.sync({
  src: { files, selectedFile },
  xform: tx.map(src => [
    "ul",
    tx.iterator(
      tx.comp(
        tx.take(5),
        tx.map(fl => [
          "li",
          {
            onclick: () => selectedFile.next(fl),
            style: {
              "font-weight": fl == src.selectedFile.name ? "bold" : "normal"
            }
          },
          fl.name
        ])
      ),
      src.files
    )
  ])
});
```

fileBrowser won't return until `selectedFile` got a `next` value. Usually, like in the commit-table-ssr example, one could just run `selectedFile.next(null)`. But here the problem is that several transformations subscribe to `selectedFile` that would then need logic to terminate early in order not to error if the value isn't a file. That's not optimal. However, if `selectedFile` is changed by another component, it's still desirable for it to update `fileBrowser`.

So I'm wondering if `sync` could have `req` and `opt` (names inspired by clojure.spec) sources, where the first result is emitted when all `req` sources returned a value and following results whenever any of the values returns a value.

**Comment by postspectacular** [comment]
Hi @den1k - it's quite funny how we seemingly encounter & work on related issues, but as always am grateful for your feedback & comments! For my own purposes I've added a `mergeOnly` option to both the `partitionSync` transducer and the [`sync`](https://github.com/thi-ng/umbrella/blob/master/packages/rstream/src/stream-sync.ts#L24) stream combinator to optionally change its behavior to simply collect & merge inputs into tuple objects and emit a new one with **each** changed input (i.e. without waiting for any others)...

To achieve your "required" input constraint, you then simply need to attach a filter to this `sync` instance to ensure the presence of the required keys:

```ts
const src = {
    a: rs.stream(),
    b: rs.stream(),
    c: rs.stream()
};

rs.sync({
    src,
    mergeOnly: true
}).transform(
    // ensure `a` & `b` are present
    // (might add this as a new transducer, e.g. `ensureKeys`...)
    tx.filter((tuple)=> tx.every((k) => x.hasOwnProperty(k), ["a", "b"]))
).subscribe(
    rs.trace("result")
);

src.c.next(1);
src.b.next(2);
src.a.next(3);
// result { c: 1, b: 2, a: 3 }
src.a.next(4);
// result { c: 1, b: 2, a: 4 }
```

**Comment by postspectacular** [comment]
Thinking further, we could add a `required` inputs option for `sync` and internalize this `filter` by composing it with the existing `partitionSync` transducer used inside `sync`... However, since `sync` also supports dynamically adding/removing inputs, I'll have to think more about how this could be achieved cleanly and not cause death-by-configuration complexity... So my current thoughts are that this composes more cleanly via the above approach.

**Comment by den1k** [comment]
re: an API for sync, here's another idea: keep `src` as is (implicitly required), and add a similar `srcOpt` object which has it's results merged into the stream value
.

**Comment by postspectacular** [comment]
Yeah, that could maybe work, though as I mentioned, the larger issue is how to deal with dynamic adds/removals of inputs (also a kind of key feature of `sync`). It's already quite complex (due to renaming sources) and this would need updating to differentiate between required vs optional as well. Will have to revisit the `partitionSync` transducer once more (which provides the main logic) and see how to feasible and/or intrusive it is to add optional keys...
